### PR TITLE
Sharathkrml/wyvern v3

### PR DESCRIPTION
**Pull request by sharathkrml**

### CODE REVIEW

File: wyvern-exchange/src/wyvern-exchange.ts [hunk 1]
```diff
@@ -1,231 +1,367 @@
...
+        }
+        let feeRecipient = royaltyDetails.feeRecipient
+        let totalRevenueETH = royaltyDetails.totalRevenueETH
+        if (from == Address.zero() || to == Address.zero()) {
+            log.error("from or to is zero, from {} to {} ,txHash {}", [
+                from.toHexString(),
+                to.toHexString(),
+                txHash,
+            ])
+            throw new Error("")
```
Comment (rahul7668gupta) [inline wyvern-exchange/src/wyvern-exchange.ts:203]: please dont throw empty errors

**Review by rahul7668gupta — CHANGES_REQUESTED** [review]

File: wyvern-exchange/src/wyvern-exchange.ts (hunk 1 above)
Comment (sharathkrml) [reply wyvern-exchange/src/wyvern-exchange.ts]: done

File: wyvern-exchange/src/wyvern-exchange.ts [hunk 2]
```diff
@@ -1,231 +1,367 @@
...
+            log.debug("from {} to {} method {} ", [
+                transfer.from.toHexString(),
+                transfer.to.toHexString(),
+                transfer.method,
+            ])
+            if (from == Address.zero()) {
+                from = transfer.from
+            } else if (from != transfer.from) {
+                log.error("from address change not expected,txhash {}", [txHash])
+                throw new Error("")
```
[the remark this answers is not in the document]
Comment (sharathkrml) [reply wyvern-exchange/src/wyvern-exchange.ts]: done

File: wyvern-exchange/src/wyvern-exchange.ts [hunk 3]
```diff
@@ -1,231 +1,367 @@
...
+                log.error("from address change not expected,txhash {}", [txHash])
+                throw new Error("from address change not expected")
+            }
+            if (to == Address.zero()) {
+                to = transfer.to
+            } else if (to != transfer.to) {
+                log.error("to address change not expected,txhash {}", [txHash])
+                throw new Error("to address change not expected")
+            }
+            if (transfer.tokens.length > 1) {
```
Comment (0xsarvesh) [inline wyvern-exchange/src/wyvern-exchange.ts:161]: Need to understand this conidition

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: wyvern-exchange/src/abi.ts [hunk 4]
```diff
@@ -302,72 +337,261 @@ export namespace abi {
...
+            let to = decoded[1].toAddress()
+            let tokenId = decoded[2].toBigInt()
+            log.debug("checkSafeTransferFrom txhash {} from {} to {} tokenId {} ", [
+                txHash,
+                from.toHexString(),
+                to.toHexString(),
+                tokenId.toString(),
+            ])
+            let tokens = new Array<Decoded_Token>()
+            let token = new Decoded_Token(Address.zero(), tokenId, BIGINT_ONE)
```
[the remark this answers is not in the document]
Comment (sharathkrml) [reply wyvern-exchange/src/abi.ts:503]: here from decoding we're not getting contract address,so we're filling this from event params

File: wyvern-exchange/src/wyvern-exchange.ts (hunk 3 above)
Comment (sharathkrml) [reply wyvern-exchange/src/wyvern-exchange.ts:161]: added for verification,this never got triggered,removed the loop below

**Pull request merged**